### DISCUSSION TITLE
Test a command that uses state

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [tauri-apps/tauri#11717] [Q&A] [answered]
I have a command in Rust that looks like the following:

```rust
pub struct AppContext {
    pub response: String,
}

#[tauri::command]
pub fn ping(state: State<'a, AppContext>) -> Result<String, String> {
    Ok(state.response)
}
```

I'd like to write a unit test for this command. Something that looks like the following:

```rust
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_pong() {
        let context = AppContext { response: "pong".into() };
        let state = State::new(context); // State's implementation in Tauri is private, so State::new() does not actually exist
        let response = ping(state).unwrap();
        assert_eq!(response, "pong".to_string());
    }
}
```

From what I can tell looking through Tauri's code is that `State`'s fields are private (so you can't just do `State(context)`) and everything that creates a `State` is private.

So it seems like it's impossible to write a unit test for commands?

--------

FWIW, with the `"test"` feature I know it's possible to instead use a `mock_builder` to instead mock the entire app and invoke commands as shown in [this Stackoverflow post](https://stackoverflow.com/questions/77524788/writing-tests-for-tauri-commands-how-to-access-and-manage-state-in-test-environ). However that opens up a lot more unknowns, and to be honest isn't that developer friendly for high unit test coverage.

See also https://github.com/tauri-apps/tauri/discussions/8424

**Comment by FabianLars** [comment, marked as answer]
What about something like this?

```rs
#[tauri::command]
fn something(state: tauri::State<String>) -> String {
    state.inner().clone()
}

#[cfg(test)]
mod tests {
    use super::something;
    use tauri::Manager;

    #[test]
    pub fn test_something() {
        let app = tauri::test::mock_app();
        app.manage("something".to_string());
        assert_eq!(&something(app.state::<String>()), "something");
    }
}
```

I'm no expert but something like `State::new()` could be problematic because then we'd have to give it a 'static lifetime which doesn't match real behavior and would allow more things than non-static lifetimes. The lifetime in the approach i showed also behaves a bit differently i think but at least it's not static so i hope it's close enough 🤔

**Reply by Nick-Mazuk** [reply]
Oh awesome, that works perfectly. Thanks Fabian!

**Reply by FelixSelter** [reply]
@FabianLars I tried the same using mock_builder() and builder.setup but my setup is never called. Do you know why this might be the case?

```rust
let builder = mock_builder();

  builder
        .setup(|app| {
            println!("Setting up test app"); // Never called
            app.manage(Mutex::new(app_state));
            Ok(())
        })
        .invoke_handler(generate_app_handler())
        .build(mock_context(noop_assets()))
        .expect("failed to build app")
```

i figured setup is only called inside make_run_event_loop_callback if you do builder.run not builder.build. So it is only recommended to use mock_app?

Nevermind I found out that the return type of builder.build is app and we can call manage on this. Sorry for disturbing

```rust
pub fn create_test_app() -> (
    tauri::WebviewWindow<MockRuntime>,
    PooledConnection<ConnectionManager<SqliteConnection>>,
) {
    init_test_db();

    let builder = mock_builder();

    let manager = ConnectionManager::<SqliteConnection>::new(TEST_DB_URL);

    let pool = Pool::builder()
        .max_size(5)
        .test_on_check_out(true)
        .build(manager)
        .expect("Failed to create DB pool");

    let app_state = AppState {
        db_connection_pool: pool,
        mdns: None,
        mdns_monitor_task: None,
        mdns_monitor_running: Arc::new(AtomicBool::new(false)),
    };
    let sql_connection = app_state
        .db_connection_pool
        .get()
        .expect("Failed to get DB connection");

    let app = builder
        .invoke_handler(generate_app_handler())
        .build(mock_context(noop_assets()))
        .expect("failed to build app");

    app.manage(Mutex::new(app_state));

    tauri::WebviewWindowBuilder::new(&app, "main", Default::default())
        .build()
        .map(|app| (app, sql_connection))
        .unwrap()
}

```
